### PR TITLE
Fix mutaby used async function argument in closure for `needless_pass_by_ref_mut`

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rust-lang/rust-clippy/issues/11380.

The problem was that it needed to go through closures as well in async functions to correctly find the mutable usage of async function arguments.

changelog: Correctly handle mutable usage of async function arguments in closures.

r? @Centri3

### CODE REVIEW

**Review by Centri3 — COMMENTED** [review]
Just two nits. Looks good otherwise

File: clippy_lints/src/needless_pass_by_ref_mut.rs [hunk 1]
```diff
@@ -439,3 +459,30 @@ impl<'tcx> Visitor<'tcx> for FnNeedsMutVisitor<'_, 'tcx> {
         }
     }
 }
+
+struct ClosuresRetriever<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    closures: FxHashSet<LocalDefId>,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ClosuresRetriever<'a, 'tcx> {
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/needless_pass_by_ref_mut.rs:468]: Good idea!

File: clippy_lints/src/needless_pass_by_ref_mut.rs [hunk 2]
```diff
@@ -95,6 +96,30 @@ fn should_skip<'tcx>(
...
+    checked_closures: &mut FxHashSet<LocalDefId>,
+    closures: FxHashSet<LocalDefId>,
+) {
+    let hir = cx.tcx.hir();
+    for closure in closures {
+        if !checked_closures.insert(closure) {
+            continue;
+        }
+        ctx.prev_bind = None;
+        ctx.prev_move_to_closure.clear();
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/needless_pass_by_ref_mut.rs:112]: Good idea.

**Comment by GuillaumeGomez** [comment]
Applied your suggestions @Centri3.

**Comment by Centri3** [comment]
@bors r+

Thanks!

**Comment by fasterthanlime** [comment]
Still seeing this on 2023-09-21, I guess the nightly build happened before the merge? I'm always unclear on what the exact timing is there.

**Comment by Centri3** [comment]
Sync hasn't happened yet, last one was #115790. It's biweekly so it'll be a few days until it's in nightly.

**Comment by conradludgate** [comment]
Hi, this issue exists on the beta build currently. Can we get this fix backported?

**Pull request merged**